Remove ServiceManager V2 compatibility

gsteel opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | yes
| QA            | yes

### Description

- Removes SMv2 compat in the InputFilterPluginManager because it cannot possibly be installed anymore
- Replaces usage of aliases for Input|Filter|Validator plugin managers with FQCN
- Adds a psalm plugin to dev dependencies for typical Psr Container methods
- Updates locked dependencies on 7.4
- Fixes a number of psalm issues, updating baseline

gsteel committed:
Remove ServiceManager V2 compatibility

- Removes SMv2 compat in the InputFilterPluginManager because it cannot possibly be installed anymore
- Replaces usage of aliases for Input|Filter|Validator plugin managers with FQCN
- Adds a psalm plugin to dev dependencies for typical Psr Container methods
- Updates locked dependencies on 7.4
- Fixes a number of psalm issues, updating baseline

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented on docs/book/specs.md:
```diff
@@ -74,7 +75,7 @@ class MyValidatingControllerFactory implements FactoryInterface
 
         // Retrieve the InputFilterManager
-        $filters = $services->get('InputFilterManager');
+        $filters = $services->get(InputFilterPluginManager::class);
```
Update the docs so that they prefer fetching the plugin manager by FQCN

gsteel commented on psalm.xml.dist:
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0"?>
 <psalm
+    errorLevel="1"
```
Fix deprecated option since a few versions ago

gsteel commented on src/InputFilterAbstractServiceFactory.php:
```diff
@@ -60,11 +59,6 @@ public function canCreate(ContainerInterface $services, $rName)
      */
     public function canCreateServiceWithName(ServiceLocatorInterface $container, $name, $requestedName)
```
Should I possibly mark these legacy methods as deprecated also?

Ocramius replied:
Deprecation will certainly allow for more graceful removal, yes.

Ocramius commented on composer.json:
```diff
@@ -40,6 +40,7 @@
     "require-dev": {
         "laminas/laminas-coding-standard": "~2.2.1",
         "laminas/laminas-db": "^2.13.4",
+        "lctrs/psalm-psr-container-plugin": "^1.6",
```
I don't think we should add this for the sake of "winning against static analysis holes" :thinking:

Ocramius replied:
A psalm `.stub.php` file would suffice for our use-case, perhaps?

gsteel replied:
> I don't think we should add this for the sake of "winning against static analysis holes" 🤔

Is that because of the dependency burden? I frequently use this plugin in projects because it avoids assigning the service to a temporary variable along with an `assert instanceof` - Not familiar with `.stub.php` - I'll have a Google…

Ocramius replied:
Yes, adding dependencies should be done with a lot of care, since most of our work right now is playing with the dependency puzzle until it clicks together.

Having a bigger baseline is still better than having more dependencies

gsteel replied:
@Ocramius - trying to get the stub right but can't figure out how to get psalm to return instance of the given class-string: https://psalm.dev/r/93dc7fec74 - Any chance you could help me out there?

Ocramius replied:
See https://psalm.dev/r/027e9ad09e

The `T` needs to be on the method in this case, since it's not the whole container that is templated with `T`

gsteel replied:
🙏

Ocramius replied:
Just cross-linking: https://github.com/Lctrs/psalm-psr-container-plugin/issues/545

gsteel commented on src/InputFilterPluginManager.php:
```diff
@@ -71,17 +77,9 @@ class InputFilterPluginManager extends AbstractPluginManager
-     *
-     * @var bool
-     */
-    protected $shareByDefault = false;
```
Perhaps dropping this is a BC break. Should I revert and mark the class as @\final?

gsteel commented on src/InputFilterPluginManagerFactory.php:
```diff
@@ -2,28 +2,39 @@
 use function is_array;
 
+/**
+ * @link ServiceManager
```
Prevents automated removal of unused use by phpcbf

Ocramius commented on src/InputFilterPluginManager.php:
```diff
@@ -71,17 +77,9 @@ class InputFilterPluginManager extends AbstractPluginManager
-     *     ContainerInterface is expected.
-     * @param array $v3config Optional configuration array (laminas-servicemanager v3 only)
+     * @param null|ConfigInterface|ContainerInterface|PsrContainerInterface $configOrContainer
+     * @param array $v3config
```
Can be removed

Ocramius commented on test/InputFilterPluginManagerCompatibilityTest.php:
```diff
@@ -31,4 +31,10 @@ protected function getInstanceOf()
+    public function testShareByDefaultAndSharedByDefault(): void
+    {
+        // The plugin manager no longer supports V2 so this inherited test is forced to pass
+        self::assertTrue(true);
```
This shouldn't be added as-is :D

gsteel committed:
Replace psalm psr-container plugin with a local stub

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Baseline method signature mismatch caused by `ServiceManager::has()`

Psalm error reads: Method Laminas\ServiceManager\ServiceManager::has with return type '' is different to return type 'bool' of inherited method Psr\Container\ContainerInterface::has - This is possibly due to @\inheritDoc

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Revert removal of legacy property `$shareByDefault` and mark deprecated

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Remove duplicate parameter doc

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on .stub.php:
```diff
@@ -0,0 +1,19 @@
+<?php
```
I think we should document this stub a bit:

 * file name more descriptive
 * a few comments here, explaining our intent

Ocramius commented on .stub.php:
```diff
@@ -0,0 +1,19 @@
+
+    interface ContainerInterface
+    {
+        /** @param string|class-string $name */
```
In theory, we can drop this declaration completely, since the original class declaration should take over :thinking: 

Can you give it a try? Let's keep all of this very minimal

gsteel replied:
It doesn't appear the declaration can be dropped - looks like the stub replaces the interface entirely.

Ocramius commented on test/InputTest.php:
```diff
@@ -48,14 +48,26 @@ public function assertRequiredValidationErrorMessage(Input $input, string $messa
-        $this->assertIsArray($messages, $message . ' non-array messages array');
+        $expectedKey = NotEmptyValidator::IS_EMPTY;
+        $messages    = $input->getMessages();
+        /** @psalm-suppress RedundantConditionGivenDocblockType */
```
We can drop this assertion overall, no?

gsteel committed:
Document the stub purpose and provide a more descriptive name

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Deprecate v2 compatibility methods

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Drop unnecessary assertion

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius reviewed (approved):
Thanks @gsteel!
